### PR TITLE
feat: language switch with multisite support

### PR DESCRIPTION
**Pull request by MaxKless**
## PR Type

[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

## What Is the Current Behavior?
Currently, the language switch only changes the language and completely ignores any multisite config.

## What Is the New Behavior?

Now the language switch works with our new default multi site setup and can be configured/overridden to work with arbitrary multi site configurations.

## Does this PR Introduce a Breaking Change?

[ ] Yes
[ x No

### CODE REVIEW

**Review by shauke — CHANGES_REQUESTED** [review]
The generated links in the language switch component look like this now for our standard multi-site configuration: `http://localhost:4200/de/home;redirect=1;lang=de_DE`
I would have expected `http://localhost:4200/de/home`. What is the reason for keeping the `;redirect=1;lang=de_DE` part. Isn't this still doing things that would no longer be necessary?

**Comment by shauke** [comment]
Should the language switch work with such a multi-site configuration as well?
```
      MULTI_CHANNEL: |
        .+:
          - baseHref: /de
            channel: default
            lang: de_DE
          - baseHref: /fr
            channel: default
            lang: fr_FR
          - baseHref: /
            channel: default
            lang: en_US
```
This currently does not work as expected.

There also might be a problem with our current `\nginx\channel.conf.tmpl`.
The part
```
  {{- if (has ($mapping | jsonpath "$..baseHref") "/") }}
  {{- else }}
    location / {
        allow all;
        auth_basic off;
        {{ $first := index $mapping 0 -}}
        rewrite ^/$ "$scheme://$http_host{{ $first.baseHref }}/home" permanent;
        rewrite ^(.*)$ "$scheme://$http_host{{ $first.baseHref }}$request_uri" permanent;
    }
  {{- end -}}
```
does not result in the same behavior as completly removing this part for the above mentioned `MULTI_CHANNEL` configuration. This might be relevant for testing.

File: src/environments/environment.model.ts [hunk 1]
```diff
@@ -138,6 +140,13 @@ export const ENVIRONMENT_DEFAULTS: Environment = {
     { lang: 'de_DE', currency: 'EUR', value: 'de', displayName: 'German', displayLong: 'German (Germany)' },
     { lang: 'fr_FR', currency: 'EUR', value: 'fr', displayName: 'French', displayLong: 'French (France)' },
   ],
+  // tslint:disable: use-camel-case-environment-properties
```
Comment (dhhyi) [inline src/environments/environment.model.ts:143]: instead of disabling this rule, adapt it 😉 https://github.com/intershop/intershop-pwa/blob/6a80bb74be4b07e49902bf6a1cd33cc4a783dd01/tslint-rules/src/useCamelCaseEnvironmentPropertiesRule.ts#L13

**Review by dhhyi — DISMISSED** [review]
I think the `multiSiteLocaleMap` must not only be a `environment.ts` property, but an environment property. It should also be possible to define complete URLs to switch to so that this feature also works when the multi-site setup is using different domains for each channel.

File: src/app/shell/header/language-switch/language-switch.component.html [hunk 2]
```diff
@@ -17,7 +17,7 @@
         <ul *ngIf="availableLocales.length">
           <ng-container *ngFor="let l of availableLocales">
             <li *ngIf="l.lang !== locale.lang">
-              <a [href]="location | makeHref: { redirect: '1', lang: l.lang }"> {{ l.displayName }} </a>
+              <a [href]="location | makeHref: { redirect: '1', lang: l.lang } | async"> {{ l.displayName }} </a>
```
Comment (shauke) [inline src/app/shell/header/language-switch/language-switch.component.html:20]: I guess the `redirect` functionality is not really needed here.
Even if the previous behavior with an `undefined` `multiSiteLocaleMap` is used will the language parameter vanish as soon as the user routes after the language switch.

File: package-lock.json [hunk 3]
```diff
@@ -4798,6 +4798,16 @@
       "integrity": "sha512-jDctJ/IVQbZoJykoeHbhXpOlNBqGNcwXJKJog42E5HDPUwQTSdjCHdihjj0DlnheQ7blbT6dHOafNAiS8ooQKA==",
       "dev": true
     },
+    "bindings": {
```
Comment (shauke) [inline package-lock.json:4801]: The `package-lock.json` changes should probably not be part of this pull request since there where no changes to the `package.json`. So there is no obvious reason why this would be needed.

File: src/environments/environment.model.ts [hunk 4]
```diff
@@ -78,6 +79,7 @@ export interface Environment {
 
   // configuration of the available locales - hard coded for now
   locales: Locale[];
+  multiSiteLocaleMap: MultiSiteLocaleMap;
```
[the remark this answers is not in the document]
Comment (shauke) [reply src/environments/environment.model.ts:89]: An additional documentation comment would be nice at this place. This was missed for a few other new parameters so we should not continue leaving those out. ;)

**Review by shauke — CHANGES_REQUESTED** [review]

File: src/app/shell/header/language-switch/language-switch.component.spec.ts [hunk 5]
```diff
@@ -57,8 +62,8 @@ describe('Language Switch Component', () => {
     expect(element.querySelectorAll('li')).toHaveLength(2);
     expect(element.querySelectorAll('[href]')).toMatchInlineSnapshot(`
       NodeList [
-        <a href="/;redirect=1;lang=en_US"> English </a>,
-        <a href="/;redirect=1;lang=fr_FR"> Fran¢aise </a>,
+        <a href="en_US"> English </a>,
+        <a href="fr_FR"> Fran¢aise </a>,
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply src/app/shell/header/language-switch/language-switch.component.spec.ts:66]: yes

File: src/app/core/utils/multi-site/multi-site.service.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  /**
+   * returns the current url, modified to fit the locale parameter if the environment parameter "multiSiteLocaleMap" is set
+   * @param locale the locale which the new url should fit
+   * @param location the current location
+   * @returns the current url, modified to fit the locale parameter
+   */
+  getLangUpdatedUrl(locale: string, url: string, baseHref: string): Observable<string> {
+    return this.stateProperties
+      .getStateOrEnvOrDefault<Record<string, unknown>>('MULTI_SITE_LOCALE_MAP', 'multiSiteLocaleMap')
```
[the remark this answers is not in the document]
Comment (shauke) [reply src/app/core/utils/multi-site/multi-site.service.ts:22]: I added an example that not yet works. Once we get it to work we can provide a commented out example in the `docker-compose.yml` and we proved that it works. Might currently just be a problem of the String/JSON format.

**Review by shauke — CHANGES_REQUESTED** [review]

**Review by shauke — APPROVED** [review]
